### ISSUE TITLE
Should the DefaultBandwidthMeter get notified (TransferListener methods to be called) about mpd file transfers ?

### ISSUE THREAD

**Issue by harismexis** [google/ExoPlayer#9904]
Hi. I would like to ask you a question please. 

We develop & maintain a video player sdk which wraps ExoPlayer. We also have a reference app for testing our sdk. The sdk is used by our clients in their video streaming apps.

We have an issue since quite a long time and we believe we may have found the cause and also applied a potential solution, 
but we are unsure if we correctly detected the cause and also of our solution.

We would like to have your advice please for the cause of the issue.

### Issue:

The issue is that when playing Live DASH with our ref app (or client app), the player is not sticking to highest video quality even when bandwidth is more than enough (100Mbps+). The player will select the highest profile soon after playback starts, but after approximately 1 min of playback the player will switch to a lower profile and will never again climb up to the highest.

### Potential Cause:

We believe that our issue is caused because we notify the DefaultBandwidthMeter when manifest file is transferred (the TransferListener methods are called) which ExoPlayer seems not to be doing.

We debugged ExoPlayer's DefaultBandwidthMeter and we can see that the onTransferStart-End-etc methods are called when a file is transferred during playback. 

**Debugging with Google ExoPlayer demo:**
I added a live DASH asset in the media.exolist.json and played the asset with breakpoints in onTransferStart-End-etc. I observed the dataSpec of each transferred file and is never an mpd file. It's always video / audio / init / subs file but never mpd. It looks like the initial fetch & periodical update of the manifest are not triggering the DefaultBandwidthMeter, so the mpd file transfers are not included in bitrate estimation. Also I checked some existing DASH assets in the demo app. Since they are not live, I see the manifest is fetched only on beginning of playback. I could not see any mpd dataSpec in my breakpoints in onTransferStart-End-etc when playback started. 
So it looks like mpd file transfers don't trigger the TransferListener methods of DefaultBandwidthMeter, so they are not included in bitrate estimation.

**Debugging with our ref app:**
When playing same live DASH asset in our ref app, I can see that mpd files do appear when doing the same debugging I did with ExoPlayer demo. Particularly for live content that updates its manifest periodically, we can see mpd file transfers triggering the onTransferStart-End on each mpd transfer.

**Our conclusion:**
Our conclusion is that the mpd file transfers which trigger the DefaultBandwidthMeter periodically during live playback are causing the issue by affecting the bitrate estimation. So we concluded that our issue happens because in our sdk we notify the DefaultBandwidthMeter for mpd file transfers. But we are worried in case the DefaultBandwidthMeter should actually be notified for mpd files, but we just did not see it for some reason in ExoPlayer, so we might got a wrong understanding.

### Current Solution:

Based on our conclusion the solution we applied was to stop notifying the DefaultBandwidthMeter for mpd file transfers. By applying our solution the issue stopped happening and we never saw the player dropping to a lower profile, but always climbing & sticking to highest when bandwidth is fair.

Before giving more details of our solution we would like to ask you if it actually sounds reasonable. Is it correct not to notify the DefaultBandwidthMeter for mpd files ? We think that's the case in ExoPlayer, that's why we did it. 

Do you think that our conclusion is correct ? 
Is it correct that ExoPlayer is not notifying the DefaultBandwidthMeter of mpd file transfers hence excluding mpd files from bitrate estimations ?
Does it sound reasonable that our issue is caused because we notify the DefaultBandwidthMeter of mpd file transfers and the proper approach to solve the issue is to stop doing this?

Thank you
Haris Mexis

**Comment by tonihei** [comment]
> Do you think that our conclusion is correct ?

Yes, that all sounds reasonable. The manifest loads are usually very small compared to the media loads and the small transfer size exacerbates measurement inaccuracies a lot. We published a bandwidth measurement analysis a while ago where we identified loads with very small transfer sizes as one issue that causes problems when estimating bandwidth: https://docs.google.com/document/d/1e3jVkZ6nxNWgCqTNibqV8uJcKo8d597XVl3nJkY7P8c/edit#heading=h.omecbu2809cn 

> Is it correct that ExoPlayer is not notifying the DefaultBandwidthMeter of mpd file transfers hence excluding mpd files from bitrate estimations ?

Yes, because of the mentioned measurement inaccuracies and also because manifests tend to be hosted on different CDN nodes (with potentially different/better caching) and this makes the bandwidth estimation even less valuable because the measured throughout to one server where the data is already cached doesn't necessarily help to understand the available bandwidth for media that comes from another server and may not have the same level of caching. 

> Does it sound reasonable that our issue is caused because we notify the DefaultBandwidthMeter of mpd file transfers and the proper approach to solve the issue is to stop doing this?

Given you already observed that it helps to solve your problem and there is a good theoretical explanation to do it like this, I'd say it's reasonable.
